fontbe/src/metrics_and_limits.rs

```diff
@@ -31,57 +35,178 @@ struct GlyphLimits {
-        self.min_right_side_bearing = self
-            .min_right_side_bearing
-            .map(|v| min(v, right_side_bearing))
-            .or(Some(right_side_bearing));
-        self.x_max_extent = self
-            .x_max_extent
-            .map(|v| max(v, bbox.x_max))
-            .or(Some(bbox.x_max));
-        self.advance_width_max = max(self.advance_width_max, advance);
-
         match glyph {
             Glyph::Simple(simple) => {
-                let points = simple.contours().iter().map(Contour::len).sum::<usize>() as u16;
-                self.max_points = max(self.max_points, points);
-                self.max_contours = max(self.max_contours, simple.contours().len() as u16);
+                let num_points = simple.contours().iter().map(Contour::len).sum::<usize>() as u16;
+                let num_contours = simple.contours().len() as u16;
+                self.max_points = max(self.max_points, num_points);
+                self.max_contours = max(self.max_contours, num_contours);
+                self.glyph_info.insert(
+                    id,
+                    GlyphInfo {
+                        gid: id,
+                        num_points,
+                        num_contours,
+                        ..Default::default()
+                    },
+                )
             }
             Glyph::Composite(composite) => {
                 let num_components = composite.components().len() as u16;
                 self.max_component_elements = max(self.max_component_elements, num_components);
+                let components = Some(composite.components().iter().map(|c| c.glyph).collect());
+                self.glyph_info.insert(
+                    id,
+                    GlyphInfo {
+                        gid: id,
+                        components,
+                        ..Default::default()
+                    },
+                )
             }
         };
     }
+
+    fn compute_component_limits(&self) -> ComponentLimits {
+        let mut depths = HashMap::<GlyphId, u16>::new();
+        let mut pending = self
+            .glyph_info
+            .iter()
+            .filter_map(|(gid, gi)| if gi.is_component() { Some(*gid) } else { None })
+            .collect::<Vec<_>>();
+        let mut components = Vec::with_capacity(8);
+        let mut max_points = 0;
+        let mut max_contours = 0;
+        while !pending.is_empty() {
+            let size_before = pending.len();
+            pending.retain(|gid| {
+                let glyph_info = self.glyph_info.get(gid).unwrap();
+                components.clear();
+                components.extend(
+                    glyph_info
+                        .components
+                        .as_ref()
+                        .unwrap()
+                        .iter()
+                        .map(|gid| self.glyph_info.get(gid).unwrap()),
+                );
+
+                // Update max points, contours while we're here
```

Comment:
I find it strange that in a method called `compute_component_limits` you update max_points and max_contours which are about non-composite glyphs... Have you not done that already in GlyphLimits::update for simple glyphs?

Reply:
sorry stupid me, you _are_ updating max_composite_points and max_composite_contours 😅